File: backend/app/schemas/analytics.py
"""Analytics schemas for data statistics and analysis."""
from datetime import date
from typing import Optional, List
from pydantic import BaseModel, Field


class DashboardStats(BaseModel):
    """Dashboard overview statistics."""
    total_goals: int = Field(description="Total number of yearly goals")
    active_goals: int = Field(description="Number of active goals")
    completed_goals: int = Field(description="Number of completed goals")
    total_monthly_plans: int = Field(description="Total monthly plans this year")
    total_daily_plans: int = Field(description="Total daily plans this month")
    total_tasks: int = Field(description="Total tasks across all plans")
    completed_tasks: int = Field(description="Total completed tasks")
    overall_completion_rate: float = Field(description="Overall task completion rate percentage")


class GoalCategoryStats(BaseModel):
    """Statistics by goal category."""
    category: str
    count: int
    completed: int
    completion_rate: float


class YearlyStats(BaseModel):
    """Yearly statistics."""
    year: int
    total_goals: int
    goal_completion_rate: float
    category_stats: List[GoalCategoryStats]
    monthly_progress: List[dict] = Field(description="Monthly goal progress data")
    total_plans: int = Field(description="Total monthly/daily plans")
    total_tasks: int
    completed_tasks: int
    task_completion_rate: float


class MonthlyStats(BaseModel):
    """Monthly statistics."""
    year: int
    month: int
    total_plans: int = Field(description="Total monthly plans")
    total_daily_plans: int = Field(description="Total daily plans in this month")
    total_tasks: int
    completed_tasks: int
    task_completion_rate: float
    daily_completion_data: List[dict] = Field(description="Daily task completion data")
    priority_distribution: List[dict] = Field(description="Task distribution by priority")
    weekly_comparison: List[dict] = Field(description="Week by week comparison")


class CompletionRateTrend(BaseModel):
    """Completion rate trend over time."""
    period: str = Field(description="Time period (daily, weekly, monthly)")
    start_date: date
    end_date: date
    data: List[dict] = Field(description="Array of {date, rate} objects")
    average_rate: float
    trend: str = Field(description="up, down, stable")


class HeatmapData(BaseModel):
    """Heatmap data for visualization."""
    start_date: date
    end_date: date
    data: List[dict] = Field(description="Array of {date, value, label} objects")


class AnalyticsResponse(BaseModel):
    """Generic analytics response wrapper."""
    success: bool = True
    data: dict

File: frontend/src/types/analytics.ts
/** Dashboard statistics */
export interface DashboardStats {
  total_goals: number;
  active_goals: number;
  completed_goals: number;
  total_monthly_plans: number;
  total_daily_plans: number;
  total_tasks: number;
  completed_tasks: number;
  overall_completion_rate: number;
}

/** Goal category statistics */
export interface GoalCategoryStats {
  category: string;
  count: number;
  completed: number;
  completion_rate: number;
}

/** Monthly progress data point */
export interface MonthlyProgressData {
  month: number;
  total: number;
  completed: number;
}

/** Yearly statistics */
export interface YearlyStats {
  year: number;
  total_goals: number;
  goal_completion_rate: number;
  category_stats: GoalCategoryStats[];
  monthly_progress: MonthlyProgressData[];
  total_plans: number;
  total_tasks: number;
  completed_tasks: number;
  task_completion_rate: number;
}

/** Daily completion data point */
export interface DailyCompletionData {
  day: number;
  total: number;
  completed: number;
  rate: number;
}

/** Priority distribution data point */
export interface PriorityDistribution {
  priority: string;
  count: number;
}

/** Weekly comparison data point */
export interface WeeklyComparison {
  week: number;
  total: number;
  completed: number;
  rate: number;
}

/** Monthly statistics */
export interface MonthlyStats {
  year: number;
  month: number;
  total_plans: number;
  total_daily_plans: number;
  total_tasks: number;
  completed_tasks: number;
  task_completion_rate: number;
  daily_completion_data: DailyCompletionData[];
  priority_distribution: PriorityDistribution[];
  weekly_comparison: WeeklyComparison[];
}

/** Completion rate trend data point */
export interface CompletionRateDataPoint {
  date?: string;
  week?: number;
  month?: number;
  year?: number;
  start_date?: string;
  end_date?: string;
  rate: number;
}

/** Completion rate trend */
export interface CompletionRateTrend {
  period: string;
  start_date: string;
  end_date: string;
  data: CompletionRateDataPoint[];
  average_rate: number;
  trend: "up" | "down" | "stable";
}

/** Heatmap data point */
export interface HeatmapDataPoint {
  date: string;
  value: number;
  level: "none" | "low" | "medium" | "high";
  total: number;
  completed: number;
}

/** Heatmap data */
export interface HeatmapData {
  start_date: string;
  end_date: string;
  data: HeatmapDataPoint[];
}
